espnet2/gan_svs/post_frontend/s3prl.py

```diff
@@ -0,0 +1,128 @@
+
+        if isinstance(fs, str):
+            fs = humanfriendly.parse_size(fs)
+        if fs != 16000:
+            logging.warning(
+                "All the upstream models in S3PRL now only support 16 kHz audio."
+            )
+
+        if download_dir is not None:
+            s3prl.util.download.set_dir(download_dir)
+
+        assert frontend_conf.get("upstream", None) in S3PRLUpstream.available_names()
+        upstream = S3PRLUpstream(
+            frontend_conf.get("upstream"),
+            path_or_url=frontend_conf.get("path_or_url", None),
+            normalize=frontend_conf.get("normalize", False),
+            extra_conf=frontend_conf.get("extra_conf", None),
+        )
+        if getattr(upstream.upstream, "model", None):
+            if getattr(upstream.upstream.model, "feature_grad_mult", None) is not None:
+                upstream.upstream.model.feature_grad_mult = 1.0
+        upstream.eval()
+
+        if layer != -1:
+            layer_selections = [layer]
+            assert (
+                not multilayer_feature
+            ), "multilayer feature will be deactivated, when specific layer used"
+        else:
+            layer_selections = None
+        featurizer = Featurizer(upstream, layer_selections=layer_selections)
+
+        self.multilayer_feature = multilayer_feature
+        self.layer = layer
+        self.upstream, self.featurizer = upstream, featurizer
+        self.pretrained_params = copy.deepcopy(self.upstream.state_dict())
+        self.frontend_type = "s3prl"
+        self.hop_length = self.featurizer.downsample_rate
+        self.tile_factor = frontend_conf.get("tile_factor", 1)
+        self.resampler = torchaudio.transforms.Resample(orig_freq=input_fs, new_freq=fs)
+        self.fs = fs
+        self.input_fs = input_fs
+
+    def _tile_representations(self, feature):
+        """Tile up the representations by `tile_factor`.
+
+        Input - sequence of representations
+                shape: (batch_size, seq_len, feature_dim)
+
+        Output - sequence of tiled representations
+                 shape: (batch_size, seq_len * factor, feature_dim)
+        """
+        assert (
+            len(feature.shape) == 3
+        ), "Input argument `feature` has invalid shape: {}".format(feature.shape)
+        tiled_feature = feature.repeat(1, 1, self.tile_factor)
+        tiled_feature = tiled_feature.reshape(
+            feature.size(0), feature.size(1) * self.tile_factor, feature.size(2)
+        )
+        return tiled_feature
+
+    def output_size(self) -> int:
+        return self.featurizer.output_size
+
+    def forward(
+        self, input: torch.Tensor, input_lengths: torch.Tensor
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        if self.fs != self.input_fs:
+            input = self.resampler(input)
+            input_lengths = input_lengths * self.fs / self.input_fs
+            input_lengths = input_lengths.ceil().long()
+
+        with torch.no_grad():
```

Comment:
Maybe instead of hard-code, we can also consider options with fine-tuning the upstream?